### PR TITLE
internal/provisioner: expose the metrics port for envoy

### LINKED ISSUE #5232: [provisioner] envoy should expose the metrics port.
**What steps did you take and what happened:**

When deploying Envoy with the contour provisioner, the generated pods expose a single port:

```yaml
    ports:
    - containerPort: 8080
      name: http
      protocol: TCP
```

It should also expose the metric port, as it is required to monitor the envoy pods using a `PodMonitor` or a `ServiceMonitor`.

```yaml
    ports:
    - containerPort: 8080
      name: http
      protocol: TCP
    - containerPort: 8002
      name: metrics
      protocol: TCP
```

The port number should be derived from `ContourDeployment.spec.runtimeSettings.envoy.metrics.port`

**What did you expect to happen:**

Deploy envoy using contour provisioner and gateway, and be able to monitor Envoy using a PodMonitor, which required that the monitoring port be exposed by the target pod.

**Environment:**

- Contour version: 1.24.2

### PR DESCRIPTION
**Pull request by izturn**
Expose the default/changed metric port for envoy

Closes #5232.

Signed-off-by: Gang Liu [gang.liu@daocloud.io](mailto:gang.liu@daocloud.io)

### CODE REVIEW

**Comment by izturn** [comment]
ping @sunjayBhatia @skriss

**Comment by Jean-Daniel** [comment]
Is the usage of --stats-port flag required ?
The last time I tried, I saw that contour was directly using the value from its `contourConfiguration` to configure the envoy metric listener. 

So I don't think there is a need to add an extra flag to the contour deployment.

File: internal/provisioner/controller/gateway.go [hunk 1]
```diff
@@ -236,6 +236,20 @@ func (r *gatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+			contourModel.Spec.RuntimeSettings.Envoy.Metrics.Port != 0 {
+
+			port := contourModel.Spec.RuntimeSettings.Envoy.Metrics.Port
+			contourModel.Spec.NetworkPublishing.Envoy.Ports = append(
+				contourModel.Spec.NetworkPublishing.Envoy.Ports,
+				model.Port{
+					Name:          "metrics",
+					ServicePort:   int32(port),
+					ContainerPort: int32(port),
+				})
```
Comment (skriss) [inline internal/provisioner/controller/gateway.go:251]: This has the effect of exposing the metrics port on the Service, which we don't do when using the default port and I don't think we want to do for custom ports. Were you intentionally trying to expose it on the Service? If so, could explain the thinking further

**Comment by izturn** [comment]
ok, i will upgrade it

**Comment by izturn** [comment]
gentle ping  @skriss @sunjayBhatia

File: internal/provisioner/objects/deployment/deployment.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ func DesiredDeployment(contour *model.Contour, image string) *appsv1.Deployment
 		switch {
 		case port.Name == "http" && port.ContainerPort != objects.EnvoyInsecureContainerPort:
 			args = append(args, fmt.Sprintf("--envoy-service-http-port=%d", port.ContainerPort))
+
```
Comment (skriss) [inline internal/provisioner/objects/deployment/deployment.go:108]: Let's revert the no-op changes to deployment.go and deployment_test.go, for clarity in the commit.

File: internal/provisioner/objects/deployment/deployment_test.go [hunk 3]
```diff
@@ -152,7 +152,8 @@ func TestDesiredDeployment(t *testing.T) {
 		"key": "value",
 	}
 
-	// Use non-default container ports to test that --envoy-service-http(s)-port
+	// Use non-default container ports to test that
```
Comment (skriss) [inline internal/provisioner/objects/deployment/deployment_test.go:155]: Let's revert the no-op changes to deployment.go and deployment_test.go, for clarity in the commit.

**Review by skriss — CHANGES_REQUESTED** [review]
Couple nits, but otherwise looks reasonable 👍

**Comment by skriss** [comment]
Please add a changelog file as well.

File: changelogs/unreleased/5101-izturn-small.md [hunk 4]
```diff
@@ -0,0 +1 @@
+Expose the default/changed metric port for envoy
```
Comment (skriss) [inline changelogs/unreleased/5101-izturn-small.md:1]:
```suggestion
Gateway provisioner: add a container port to the Envoy daemonset/deployment for the metrics port.
```

File: internal/provisioner/objects/dataplane/dataplane.go [hunk 5]
```diff
@@ -136,14 +136,32 @@ func desiredContainers(contour *model.Contour, contourImage, envoyImage string)
 		ports = append(ports, p)
 	}
 
-	healthPort := 8002
+	var (
+		healthPort  = 8002
```
Comment (skriss) [inline internal/provisioner/objects/dataplane/dataplane.go:140]: while we're here, it'd be nice to move this constant into `object.go` along with the other port numbers, similar to `EnvoyMetricsPort`.

**Review by skriss — CHANGES_REQUESTED** [review]

**Review by skriss — APPROVED** [review]
LGTM, thanks @izturn

**Pull request merged**